### PR TITLE
(#1) initial implementation of new streaming JSON writer

### PR DESCRIPTION
**Pull request by eli-darkly**
This is the first half of the proposed new streaming JSON library for use in the LaunchDarkly Go SDK and other LaunchDarkly components.

There will be two parts to this library: a JSON writer and a JSON reader. This PR is for the writer.

Go's standard `json.Marshal` function has a number of problems from an efficiency standpoint. One is the overhead of using reflection to inspect types. Another is that if you implement a custom `MarshalJSON` method for any type, it has to return an intermediate chunk of output as a byte slice, which means that every little piece has to be allocated separately on the heap before being appended to the output buffer. Letting marshalers push data into a stream instead can be much more efficient.

We previously implemented a streaming JSON writer as part of [go-sdk-common](https://github.com/launchdarkly/go-sdk-common/tree/v2/jsonstream), mostly to improve efficiency in the SDK when writing analytics events. There were several design problems with that code, so this version differs from it as follows:

* The `Writer` API has been considerably improved to make it easier to write custom marshaler code (see below).
* In some of our internal components, we use [easyjson](https://github.com/mailru/easyjson) to generate fast custom marshalers for various types. Easyjson already has a streaming output API, so if we're already using that to marshal type A which contains type B, and we have a marshaler for type B that uses this new library, it's desirable to have it write directly into the easyjson stream rather than creating intermediate byte arrays. This is now very easy to do: if you add the build tag `launchdarkly_easyjson` when building your code, my default implementation of JSON writing is replaced by an implementation that delegates to easyjson.
* There are much more thorough unit tests, which generate thousands of permutations of JSON data and run both the default implementation and the easyjson implementation against these to verify that they behave consistently.

Here's an example of what a custom marshaler would look like that uses this API.

```go
type MyStruct struct {
    StringField string
    ArrayOfIntsField []int
}

func (m MyStruct) WriteToJSONWriter(w *jwriter.Writer) {
    obj := w.Object()
    obj.String("stringField", m.StringField)
    obj.Property("arrayOfIntsField")
    arr := obj.Array()
    for _, n := range m.ArrayOfIntsField {
        arr.Int(n)
    }
    arr.End()
    obj.End()
}
```

It's very easy to implement a `MarshalJSON` method so the same marshalling logic can be reused if `MyStruct` is serialized with `json.Marshal`:

```go
func (m MyStruct) MarshalJSON() ([]byte, error) {
    return jwriter.MarshalJSONWithWriter(m)
}
```

And, in code that uses easyjson, you can also easily implement a `MarshalEasyJSON` method so that the same logic can be reused if a `MyStruct` appears within some other data structure that's being marshaled with easyjson. (Note that here the `ejwriter` package refers to easyjson's `jwriter`.)

```go
func (m MyStruct) MarshalEasyJSON(ejw *ejwriter.Writer) {
    w := jwriter.NewWriterFromEasyJSONWriter(ejw)
    m.WriteToJSONWriter(&w)
}
```

Like the earlier version in go-sdk-common, this also lets you configure a writer that dumps its output to an `io.Writer` at intervals instead of buffering the whole thing in memory. This is mainly intended for writing large data structures to HTTP requests or responses, as when we send event data from the SDK.

### CODE REVIEW

File: internal/commontest/interfaces.go [hunk 1]
```diff
@@ -0,0 +1,65 @@
+package commontest
+
```
Comment (eli-darkly) [inline internal/commontest/interfaces.go:2]: This and the rest of the commontest package provide a framework for running standardized test suites with a wide variety of JSON data. For instance, both the default implementation and the easyjson implementation of the low-level token writer have methods for writing nulls, bools, strings, etc.; so, we want to make sure the tests cover nulls, true bools, false bools, strings with or without escape characters, and so on, and that we test the same values with both implementations. Then there's the higher-level `Writer` API that wraps the token writer, and we'll run the same tests against that too— plus various permutations of writing arrays and objects that contain all of those kinds of values. This ensures that we don't miss some edge case where for instance it fails if you happen to write a bool right after a string.

When I add the reader package, we'll reuse basically this same logic to test the reader as well.

File: jwriter/token_writer_default.go [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+package jwriter
+
+import (
+	"encoding/json"
+	"io"
+	"strconv"
+	"unicode/utf8"
+)
+
```
Comment (eli-darkly) [inline jwriter/token_writer_default.go:11]: The logic in this file is nearly identical to the earlier implementation in go-sdk-common. There are only so many ways you can write JSON output. The `streamableBuffer` type, which provides the ability to pipe the output to an io.Writer, is identical to the version in go-sdk-common.

File: jwriter/package_example_test.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package jwriter
+
+import "fmt"
+
+func Example() {
```
Comment (hroederld) [inline jwriter/package_example_test.go:5]: I'm not sure exactly what this file is for?

File: jwriter/token_writer_default.go [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+var (
+	tokenNull  = []byte("null")  //nolint:gochecknoglobals
+	tokenTrue  = []byte("true")  //nolint:gochecknoglobals
+	tokenFalse = []byte("false") //nolint:gochecknoglobals
+)
+
+type tokenWriter struct {
+	buf       streamableBuffer
+	tempBytes [50]byte
```
Comment (hroederld) [inline jwriter/token_writer_default.go:26]: Is this magic number derived from some particular formula or is it just a reasonable guess of the needed size?

Comment (eli-darkly) [reply jwriter/token_writer_default.go:26]: If you look at where it's actually used, this is for converting ints and floats to strings, and also for building hex escape sequences like `\u0000`. The only one of those where the maximum length is indeterminate is floats, so I left some excess space there, although if `strconv.AppendFloat` runs out of room it'll just allocate a new buffer.

File: jwriter/package_example_test.go (hunk 3 above)
Comment (eli-darkly) [reply jwriter/package_example_test.go:5]: https://blog.golang.org/examples

Comment (hroederld) [reply jwriter/package_example_test.go:5]: TIL

**Review by hroederld — APPROVED** [review]
